File: src/main/java/me/deecaad/weaponmechanicsplus/weapon/modifiers/ShootModifier.java
package me.deecaad.weaponmechanicsplus.weapon.modifiers;

import me.deecaad.core.file.SerializeData;
import me.deecaad.core.file.Serializer;
import me.deecaad.core.file.SerializerException;
import me.deecaad.weaponmechanics.weapon.projectile.weaponprojectile.Projectile;
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.ModifierHelper;
import org.jetbrains.annotations.NotNull;

public class ShootModifier implements Serializer<ShootModifier>, ModifierHelper {

    private Projectile overrideProjectile;

    /**
     * Default constructor for serializer.
     */
    public ShootModifier() {
    }

    public ShootModifier(Projectile overrideProjectile) {
        this.overrideProjectile = overrideProjectile;
    }

    public Projectile getOverrideProjectile() {
        return overrideProjectile;
    }

    @NotNull
    @Override
    public ShootModifier serialize(SerializeData data) throws SerializerException {
        Projectile overrideProjectile = data.of("Override_Projectile").serialize(Projectile.class);
        return new ShootModifier(overrideProjectile);
    }
}

File: src/main/kotlin/me/deecaad/weaponmechanicsplus/weapon/modifiers/ShootModifier.kt
package me.deecaad.weaponmechanicsplus.weapon.modifiers

import me.deecaad.core.file.*
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.DoubleModifier
import me.deecaad.weaponmechanicsplus.weapon.modifiers.util.IntegerModifier

class ShootModifier : Serializer<ShootModifier> {

    var projectileAmount: IntegerModifier? = null
    var projectileSpeed: DoubleModifier? = null

    /**
     * Default constructor for serializer.
     */
    constructor()

    constructor(projectileAmount: IntegerModifier?, projectileSpeed: DoubleModifier?) {
        this.projectileAmount = projectileAmount
        this.projectileSpeed = projectileSpeed
    }

    @Throws(SerializerException::class)
    override fun serialize(data: SerializeData): ShootModifier {

        val projectileAmount = data.of("Projectile_Amount").serialize(IntegerModifier::class.java)
        val projectileSpeed = data.of("Projectile_Speed").serialize(DoubleModifier::class.java)

        return ShootModifier(projectileAmount, projectileSpeed)
    }
}